#+TITLE: Coverage
#+AUTHOR: Njagi Mwaniki
#+OPTIONS: date:nil
#+OPTIONS: toc:nil

* bwa
BWA represents the more traditional aligner or alignment methods

** Mixed
#+CAPTION: Mixed perfect, realistic and deletion coverage in igv
#+NAME:   img:mixed_coverage
[[../../Images/simulation/coverage/bwa/igv_all.png]]


This is how well the reads from sequencing cover the genome
Most of this follows The Biostar Handbook

We shall use the ebola genome.
Assuming you have efetch from [[../../Tools/Data/Entrez-Direct.org#fetch][Entrez-Direct under fetch]]
#+BEGIN_SRC
mkdir -p refs
efetch -db=nuccore -format=fasta -id=AF086833 > refs/reference.fa 

# index the reference for alignment
bwa index refs/reference.fa

# make it easy to find the ref
export MAIN_REF=<path/to/ref>
#+END_SRC

** Perfect data

There is no consensus on perfect data but let's go with what the Biostar handbook provides.
After all our aim is to compare these tools with graph tools.

#+BEGIN_SRC
mkdir perfect_coverage
cd perfect_coverage

# Get the script used in the biostar handbook
curl -O http://data.biostarhandbook.com/align/perfect_coverage.py

# Generate perfect data.
cat $MAIN_REF | python perfect_coverage.py

# Generate an alignment
bwa mem $MAIN_REF R1.fq R2.fq | samtools sort > perfect.bam 

# Index the alignment
samtools index perfect.bam
#+END_SRC

Open perfect.bam in igv to get

#+CAPTION: Perfect coverage
#+NAME:   img:perfect_coverage
[[../../Images/simulation/coverage/bwa/igv_perfect.png]]

** Realistic data
For realistic data we shall simulate reads and some mutations in the reference using  [[../../Tools/Simulation/README.org][wgsim]].

#+BEGIN_SRC
# We will make a copy of our reference file. We will call this the genome.fa
# and we will use it as the source of our simulations.
# We will modify our genome file, 

mkdir realistic_coverage
cd realistic_coverage

cp $MAIN_REF genome.fa

# Generate 10,000 reads from it
wgsim -N 10000 genome.fa r1.fq r2.fq

# Compare it to the original reference
bwa mem $MAIN_REF r1.fq r2.fq | samtools sort > realistic.bam 

# Index the alignment
samtools index realistic.bam
#+END_SRC

Open realistic.bam in igv to get

#+CAPTION: Realistic data
#+NAME:   img:perfect_coverage
[[../../Images/simulation/coverage/bwa/igv_realistic.png]]

** Deletion
In this case we make a deletion in the reference from which we will generate our reads
Get [[https://replikation.github.io/bioinformatics_side/tools/emboss/][seqret from eboss]]  described under [[../../Tools/Data/Emboss.org][Emboss]] to delete part of the genome or delete a chunk manually.

#+BEGIN_SRC
mkdir deletion
cd deletion

# Create a reference with a deletion
cat $REF | seqret --filter -sbegin 1 -send 2000 > part1
cat $REF | seqret --filter -sbegin 3000 > part2
cat part1 part2 | union -filter > genome.fa

# Generate 10,000 reads from it
wgsim -N 10000 genome.fa r1.fq r2.fq

bwa mem $REF r1.fq r2.fq | samtools sort > deletion.bam

samtools index deletion.bam
#+END_SRC

Open deletion.bam in igv to get

#+CAPTION: Coverage of a deletion in the reference
#+NAME:   img:deletion
[[../../Images/simulation/coverage/bwa/igv_deletion.png]]

* bcalm
** Perfect data
bcalm gives a single node graph for perfect data despite the kmer size and abundance estimates

#+CAPTION: bcalm gfa with kmer size of 31 and minimum kmer abundance of 20
#+NAME:   img:mixed_coverage
[[../../Images/simulation/coverage/bcalm/perfect_coverage_k31_a20.png]]

** Realistic data

Here we have increasing minimum abundance but a constant kmer size of 31

Low minimum abundance of 2
This yields a more linear graph that is in line with what we expected

[[../../Images/simulation/coverage/bcalm/realistic_k31_a2.png]]

Higher minimum abundance of 10
Some unitigs align but others don't

[[../../Images/simulation/coverage/bcalm/realistic_k31_a10.png]]

Higher minimum abundance of 20
With a higher min abundance we get unitigs that don't align

[[../../Images/simulation/coverage/bcalm/realistic_k31_a20.png]]
