#+TITLE: Reference pangenome graph
#+SUBTITLE: How to build the reference pangenome graph
#+AUTHOR: Njagi Mwaniki
#+OPTIONS: date:nil

Generate a sufficiently simple (de Bruijn) graph using an assembler suited to
the data. Let's use the simulated data from [[./Data.org][the previous data step]].

* Pangenome graph generation
Tools that can build a de Bruijn graph from reads


** spades
Spades has output in ~gfa1~ and ~fastg~.
 
#+BEGIN_SRC
spades.py --isolate -k 31 \
  -1 ~/projects/Masters/data/refs/r1.fq \
  -2 ~/projects/Masters/data/refs/r2.fq \
  -o ~/projects/Masters/results/spades
#+END_SRC

When the gfa is visualized it's linear as below and as expected:

[[../../Images/Overall/Reference/spades_realistic_simulation.png]]


** minia3

Using https://github.com/GATB/gatb-minia-pipeline

I had to install pysam via conda
~~

#+BEGIN_SRC
~/Software/gatb-minia-pipeline/gatb \
  --kmer-sizes 31 \
  -1 ~/projects/Masters/data/refs/r1.fq \
  -2 ~/projects/Masters/data/refs/r2.fq \
  -o WHUCoV 
#+END_SRC

I had problems installing pysam to I used 

#+BEGIN_SRC
~/Software/bcalm/scripts/convertToGFA.py WHUCoV_k31.contigs.fa WHUCoV_k31.contigs.gfa  31
#+END_SRC

[[../../Images/Overall/Reference/minia_realistic_simulation.png]]

** bifrost
#+BEGIN_SRC
Bifrost build \
 -k 31 \
  -s ~/projects/Masters/data/refs/interleaved.fq \
  -r ~/projects/Masters/data/refs/WHUCoV.fa \
  -o WHUCoV 
#+END_SRC

The graph looks similar to what we'd expect
[[../../Images/Overall/Reference/bifrost_realistic_simulated.png]]

** bcalm2
For bcalm first interleave the fastq files
#+BEGIN_SRC
interleave-fastq r1.fq r2.fq > interleaved.fq
# WHUCoV is just short for WuHang Corona Virus
Bifrost build \
 -k 31 \
  -s ~/projects/Masters/data/refs/interleaved.fq \
  -r ~/projects/Masters/data/refs/WHUCoV.fa \
  -o WHUCoV
#+END_SRC


[[../../Images/Overall/Reference/bcalm_realistic_simulation.png]]

* Parameters
** kmer size
Picking a kmer size is...
I chose a kmer size of 31 because I figured it was a good kmer size from
  - Wikipedia [[https://en.m.wikipedia.org/wiki/K-mer#Choice_of_k-mer][Choice of k-mer]]
  - [[https://twitter.com/urbanslug/status/1216718494328401921][this thread]] from Pall Melsted bcalm author
  - this excerpt from [[https://www.nature.com/articles/nbt.2023][Nature Biotech: How to apply de Bruijn graphs to genome assembly]]

/Reads of 100-mers generated by Illumina technology capture only a small/
/fraction of 100-mers from the genome (even for samples sequenced to high/
/coverage), thus violating the key assumption of de Bruijn graphs./
/However, if one breaks these reads into shorter k-mers, the resulting k-mers/
/often represent nearly all k-mers from the genome for sufficiently small k./
/For example, de Bruijn graph–based assemblers may break every 100-nucleotide/
/read into 46 overlapping 55- mers and further assemble the resulting 55-mers./
/Even if some 100-mers occurring in the genome are not generated as reads, this/
/‘read breaking’ procedure13 ensures that nearly all 55-mers appearing in the/
/genome are detected. In the example shown in Figure 3, the five reads do not/
/account for all 7-mer substrings of the genome. But they do contain all 3-mers/
/present in the genome, and this is sufficient to reconstruct the genome./



** minimum abundance

* Bluntify the reference genome graph

* Organize the graph


** odgi sort
