#+TITLE: Reference pangenome graph
#+SUBTITLE: How to build the reference pangenome graph
#+AUTHOR: Njagi Mwaniki
#+OPTIONS: date:nil

Generate a sufficiently simple (de Bruijn) graph using an assembler suited to
the data. Let's use the simulated data from [[./Data.org][the previous data step]].

* Pangenome graph generation
Tools that can build a de Bruijn graph from reads


** spades

Spades has output in ~gfa1~ and ~fastg~.
 
#+BEGIN_SRC
spades.py --isolate
  -1 ~/projects/experiments/simulation/data/realistic_coverage/r1.fq \
  -2 ~/projects/experiments/simulation/data/realistic_coverage/r2.fq \
  -o ~/projects/experiments/simulation/results/alignment/spades
#+END_SRC

Looking at ~assembly_graph_with_scaffolds.gfa~
It doesn't render continuously (made of different non connected graphs)
as seen below but runs through astarix successfully

[[../../Images/Overall/Reference/spades_realistic_simulation.png]]



** minia3

** bifrost

** bcalm2

* Parameters
** kmer size
Picking a kmer size is...

** minimum abundance

* Bluntify the reference genome graph

* Organize the graph


** odgi sort
