#+LATEX: \pagewidth
* Abstract
#+BEGIN_ABSTRACT
Reference genomes represent a consensus assembly of the most frequent reads or
bases in a sample. Genome graphs use low level data structures to provide 
access to reads used in the assembly as a network of nodes where a node 
represents a sequence or kmer. In this work we used genome graphs to explore
and represent virus sequence diversity.

Respiratory Syncytial Virus samples were collected from a twenty-five-member 
household during a household outbreak and used to assemble a genome graph. 
Reads from each sample were mapped on to this graph and the number of reads that
mapped under each was noted. We then applied principal component analysis to 
reduce the 196,488-dimensional coverage data and the results used to generate 
distance matrices which were then summarised as a cladogram.

We established a method able to differentiate each of the samples by comparing
the depth of coverage of each sample against that of every other sample.
The use of different principle components, either the first principal component 
or the first five principal components, gave different topologies.
The first five provided a tree that was more parsimonious and closer to the
maximum likelihood tree generated in previous studies.

Our na√Øve approach provides an all vs all sample comparison of sample
relatedness using data provided by a genome graph. It also shows that sequence
coverage across a genome graph is an alternative approach to provide sequence 
relatedness.
#+END_ABSTRACT
