\begin{abstract}
There are several methods that are used to evaluate and describe sequence diversity, and these include Shannon's entropy, the number of polymorphic loci per kilobase (one thousand bases), and the nucleotide diversity statistic Pi (Ï€). These methods compare nucleotide substitutions, insertions and deletions present in consensus sequences to describe and quantify sequence diversity from a given sample. However, such approaches are prone to underestimation of the actual diversity for example in cases of low abundance haplotypes given that consensus sequences are a mosaic of closely related haplotypes.

We propose to use variation graphs, that is, data structures that maintain available sequence variation from a sample or a collection of samples, to explore the sequence variation of respiratory syncytial virus (RSV), a single stranded, negative sense, enveloped RNA virus. This study will utilize sequence data from samples collected in a twenty-member household during the course of a household RSV outbreak. 

We aim to create a pipeline for constructing a variation graph for describing virus diversity during the household outbreak. We will use this data to assess the utility of this pangenome in informing potential transmission events.
\end{abstract}
