#+LATEX: \pagewidth
* Abstract
#+BEGIN_ABSTRACT
Linear reference genomes are a consensus of the most frequent base at a given
position.
Graph-based reference genomes represent a genome as a network of alternative
paths, thereby providing information on variant bases.
This graph can be node labelled—bases are held in the nodes of the graph while
the edges represent the connections between the bases, or edge labelled—if the
edges in the graph represent the bases. A graph-based data structure like this
is suitable for exploring and describing virus sequence diversity.
\todo{This last sentence can be better}

Sequenced respiratory syncytial virus raw reads from a twenty-five-member
household collected during a household outbreak were used to generate a genome
graph.
The sample reads were then aligned to the genome graph.
The number of reads that mapped under each node was then summarised as a
multidimensional coverage vector.
A pairwise Euclidean distance matrix of was computed and a neighbour-joining
cladogram based on hierarchical clustering generated.

We demonstrate the plausibility of differentiating a large number of closely
related consensus genomes by comparing the number of respective raw reads that
align to each node from the larger genome graph.
Additionally, the sequence coverage across a genome graph provides an
alternative approach for examining sequence relatedness and identifying
potential sequencing errors.
#+END_ABSTRACT
