The study set out to establish whether comparing samples based on how their
sequence reads aligned to a reference graph provides sufficient resolution to
understand who infects the infant with RSV in the household.
It found that by mapping sequence reads to a reference genome graph, we gain
high enough resolution to differentiate and cluster samples in the household.
We are therefore performing an all vs all sample comparison through a
representative proxy, coverage against a graph.
It is also worth noting that using this method, we were able to differentiate
samples better than the tools used in previous studies.
The study did not go far enough to try and establish transmission chains in the
household or the circulation of different subtypes and genotypes, but that does
not mean that it is not possible.

Contemporary methods of reference genome representation use a linear model, a
golden path, to represent a shared genome and a separate file to hold variation
information such as a VCF file leading to a failure to account for some of the
genetic variations during mapping.
This model is lossy as a result of its linearity because it is a consensus of
the most frequently occurring sequences amongst the sequences used to build it.
The lack of lower frequency variants, in turn, causes the model to favour
high-frequency variants leading to a tendency to over-report sequences present
in the reference and under-report those that are not.
This tendency, known as reference allele bias, leads to an underestimation of
the genetic variation present in a genome. Moreover, the reference is not
transitive ($A>B\ and\ B>C,\ therefore,\ A>C$) and therefore not an ideal way to compare samples.
That is, how a sample  compares to reference and how another sample  compares to
the reference does not mean that it is possible to conclusively compare  against
based on how they compare against the reference.

Graph-based models sidestep the lack of transitivity problem by being able to
present alternatives and when used to represent the reference genome have been
shown to reduce reference allele bias, consequently facilitating the comparison
of genomes at a higher resolution than before.

When it comes to who infects whom in the household; the premise of this work is
that sample relatedness is a proxy for transmission.
I expect a person to have a virus that is genetically similar to that of the
people the person infects. Therefore, infections of similar origin will cluster
together because like samples will cluster together.

We find all one but one (504) of the samples collected in March cluster together
in the clade highlighted in red while the samples collected in April cluster in
the clades annotated in blue and green. It is also fascinating to see that
member, 518, a school going eleven years old, has their samples sit on their
own, implying they did not acquire their RSV infection from a member of the
household. This is very interesting and expected because it shows that samples
collected at around the same time were genetically similar.
It also shows that the samples from the school going child were consistently
different from the others.

It is highly likely that the individual whose sample clustered closest to that
of the infant could have been the ones who transmitted the virus to the infant.
In this case, the two-year-old, 528, had their sample sit closest to that of the
infant, implying that the infant's RSV infection is from the two-year-old.

** Strengths and limitations
Genome graphs give insight into genetic variation in ways that are not possible
with existing tools.
It was possible to come up with an intuitive method, useful for comparing
samples through the aid of genome graphs which provide a higher resolution when
looking at pangenomes or in our case a virus quasispecies. By representing
alternative bases, aligning reads against these alternative bases, and then
analysing the coverage of these alignments, it is possible to analyse genomes at
a higher resolution than the methods that make use of linear references do.

This study did not attempt to differentiate between genotypes and subtypes of
RSV or investigate and how they circulate within the household.
It is also not entirely conclusive and uses a naive approach that does not
account for noise or the varying level of significance between mapping to a
given node versus another.
