* RSV
We started the experiment with fifty three whole genome sequenced RSV 
positive samples from a twenty five member household code-named household twenty
which was part of a household follow up study in the Kilifi Health and 
Demographic Surveillance System and ran over a six month
period collecting nasopharyngeal swabs twice each week
cite:munywokiInfluenceAgeSeverity2015,agotiTransmissionPatternsEvolution2017,githinjiAssessingUtilityMinority2018.

Figure [[fig:rsv_temporal_distribution]] from
cite:githinjiAssessingUtilityMinority2018 shows the temporal infection patterns
of nineteen of the twenty five members of household twenty.

#+CAPTION[Temporal Distribution of RSV Sample Collection]: The y-axis shows anonymized individuals represented by integers and unique colors and on the x-axis is the dates on which they were sampled.
#+ATTR_LATEX: :placement [h] :width 0.75\textwidth :float multicolumn
#+NAME: fig:rsv_temporal_distribution
file:../Figures/RSV/Githinji_HH_5_temporal_distribution.png

** Preprocess and Quality Control
We started by removing reads with a Phred score below 35 and trimming sequencing
adapters as shown in [[Quality Control and Adapter Trimming]] then compiled a list 
of the file paths into a text file as shown in [[Concatenate Reads into Text File]].

Sequences in two datasets (~H_536_09_04~ and ~H_506_13_04~), specifically those 
stated [[Problematic RSV Sequences]], caused seqwish
(https://github.com/ekg/seqwish) to crash and were therefore removed from the 
experiment leaving us with 51 samples.

** Construct the Assembly Graph
We used minia cite:chikhiSpaceefficientExactBruijn2013 to construct two assembly
graphs with a k-mer length of 31 and varying minimum abundance values.
Given the nature of the dataset (51 samples of approximately 400-500 megabytes 
from a genome that is approximately 15 kilobases in length) it was clear 
that the data was noisy and we had to set a high minimum abundance. 
We varied the minimum abundance values between 1,000 and 2,000 as shown 
in [[Minia Fragment Assembly]] to get a graph of size not greatly exceeding that of
the RSV genome.

A minimum abundance of 1,000 resulted in 157 kilobyte
(approximately 10x the genome size) GFA while a minimum abundance of 2,000 
resulted in a 237 kilobyte GFA (approximately 20x the genome size).
Increasing the minimum abundance from 1,000 to 2,000 reduced the size of the 
resulting graph but on visual inspection seemed to have lost some of 
the variable regions as seen in Figure [[fig:rsv_assembly]].

#+CAPTION[RSV Assembly Graph]: An assembly graph of the household 20 samples built using minia and a minimum abundance of 1000 left and 2000 to the right.
#+ATTR_LATEX: :placement [h] :scale 1.0 :float multicolumn
#+NAME: fig:rsv_assembly
file:../Figures/RSV/Assembly_Combined.png

** Bluntify the Assembly Graph
Going with the 1,000 minimum abundance graph for its increased variability but 
still manageable size we used stark (https://github.com/hnikaein/stark) to 
bluntify, that is reduce the overlaps on edges,
 cite:gargGraphbasedApproachDiploid2018 of the graph as shown in
[[Graph Bluntification Using Stark]].
This yielded a 206 kilobyte GFA that when visualized is seen in Figure
[[fig:rsv_bluntified_assembly_graph]].

** Prepare the Graph for Mapping with VG
The bash script in [[Odgi Graph Preparation]] was used to chop and sort the graph
for use with vg cite:garrisonVariationGraphToolkit2018 which led to the 
variation graph in Figure [[fig:rsv_variation_graph]].

** Mapping
*** Convert GFA to vg Compatible Variation Graph
Using the instructions in [[RSV VG View]], we induced a vg 
cite:garrisonVariationGraphToolkit2018 compatible variation graph and output it 
GFA.

*** Index
There was no need to prune the graph because it was small 
(a 62 kilobyte graph.vg) and we wanted to avoid losing complex regions such as 
those with many variants close to each other.
We, therefore, only built an index as in [[RSV VG Index]] and got a 61
kilobyte graph (graph.xg) and a 258 kilobyte index (graph.gcsa).

*** Mapping
To map each sample against the graph, we used [[RSV VG Map Script]] to loop through 
each of our interleaved sequences and stored the output GAM files in a 
directory of our specification.

We then verified that the alignments made sense
(a task that is both subjective and a matter of judgment) by converting the GAM 
to GAMP using the instructions in  [[Generate GAM]] and viewing the JSON.

** Calculate Coverage Across the Graph for Each Sample
We used the script in [[VG Pack Script]] to generate a coverage vector 196,488 nodes
long in tsv (tab separated values) format.

** Normalization
The coverage data had a lot of outliers, as shown in Figure
[[fig:structure_of_data]], and therefore needed normalization to avoid the outliers
skewing it.

#+CAPTION[RSV Structure of the Data]: Bar graphs of mean, median, maximum and standard deviation of coverage values per sample
#+ATTR_LATEX: :placement [h] :width 0.75\textwidth :float multicolumn
#+NAME: fig:structure_of_data
file:../Figures/RSV/structure_of_coverage_data.png


We normalized the coverage by setting any coverage value above zero as one and 
left the zero values as zero. This meant that in this analysis, any form of
coverage no matter how deep was valued equally which yielded the heatmap in
Figure [[fig:rsv_heatmap]].

#+CAPTION[RSV heatmap]: A heatmap of the binary normalized coverage vectors of the forty nine RSV samples. On the x axis is the node identifier and the y axis are the individual samples. The light regions indicate coverage while the dark regions indicate no coverage.
#+ATTR_LATEX: :placement [h!] :width 0.7\textwidth :float multicolumn
#+NAME: fig:rsv_heatmap
file:../Figures/RSV/Heatmap.png

#+LATEX: \newpage
** Principal Component Analysis
To make it possible to compare the high dimensional data, we applied Principal 
Component Analysis (PCA) which was able to differentiate each of the samples.
Figure [[fig:rsv_pca]] is a scatter plot of the first and second principal 
components for RSV reads.

#+CAPTION[RSV PCA]: A scatter plot of the first and second principal components of the coverage vectors of the forty nine RSV samples.
#+ATTR_LATEX: :placement [h] :width 0.75\textwidth :float multicolumn
#+NAME: fig:rsv_pca
file:../Figures/RSV/PCA.png


#+CAPTION[Bluntified RSV Assembly Graph]: RSV household 20 assembly graph bluntified using stark.
#+ATTR_LATEX: :placement [h] :width 0.75\textwidth :float multicolumn
#+NAME: fig:rsv_bluntified_assembly_graph
file:../Figures/RSV/Assembly_Bluntified.png

#+CAPTION[RSV Variation Graph]: The household 20 variation graph after running odgi chop on it.
#+ATTR_LATEX: :placement [h!] :width 0.75\textwidth :float multicolumn
#+NAME: fig:rsv_variation_graph
file:../Figures/RSV/Assembly_Bluntified.png

#+LATEX: \clearpage
