#+TITLE: bcalm
#+AUTHOR: Njagi Mwaniki
#+OPTIONS: date:nil
#+OPTIONS: toc:nil

[[https://github.com/GATB/bcalm][bcalm]] is a de Bruijn graph compaction in low memory 

* Installation
To compile follow the [[https://github.com/GATB/bcalm#installation][instructions in the README]]

In my case: guix and zlib weren't playing well so I [[https://anaconda.org/bioconda/bcalm][installed bcalm using conda]].

#+BEGIN_SRC
conda install -c bioconda bcalm
#+END_SRC

* Usage
bcalm takes a single interleaved fastq file
If data is in seperate files interleave it with: [[https://github.com/ekg/interleave-fastq][iterleave-fastq]]

To use interleave-fastq

#+BEGIN_SRC
interleave-fastq R1.fq R2.fq > interleaved.fq
#+END_SRC


#+BEGIN_SRC
bcalm -in ~/projects/Martha/HRV/data/81/clean/HRV_interleaved.fastq -kmer-size 31 -abundance-min 10
#+END_SRC

Convert to gfa using 

#+BEGIN_SRC
~/Software/bcalm/scripts/convertToGFA.py HRV_interleaved.unitigs.fa HRV_31_20.gfa 31
#+END_SRC

* Results
Gives a ~.h5~, ~.unitigs.fa~, ~.unitigs.fa.glue~ and ~.unitigs.fa.glue.<number>~

From [[https://github.com/GATB/bcalm#intermediate-files][bcalm docs intermediate files]]

"/BCALM 2 produces some intermediate files: a .h5 file (or a _gatb/ folder),/
/which contain the k-mer counts. The "*glue*" files contain compacted sequences/
/that needs to be glued together (see BCALM 2 paper). Those files can be safely/
/deleted after an execution, as the actual output is just the FASTA file/
/containing the unitigs./"


Given the actual output is the ~.unitigs.fa~, bcalm provides
[[https://github.com/GATB/bcalm/blob/master/scripts/convertToGFA.py][scripts/convertToGFA.py]] to convert unitigs.fa to .gfa

For example:
#+BEGIN_SRC
# 31 is the kmer size and the output is ~perfect_coverage.gfa~
~/Software/bcalm/scripts/convertToGFA.py perfect_coverage_interleaved.unitigs.fa perfect_coverage.gfa 31
#+END_SRC
